File: libs/vulnfounder-core/parsers/zig/repository_scanner.py
"""
Stage 1: Repository Scanner for Zig

Enumerates all Zig source files in a repository.
"""

import os
from core.repo_walk import walk_repository
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any, Optional

from utilities.file_io import write_json
from utilities.path_filters import should_exclude_directory


class RepositoryScanner:
    """Scans a repository for Zig source files."""

    # Directories to exclude from scanning
    EXCLUDE_DIRS = {
        ".git",
        "vendor",
        "node_modules",
        "zig-cache",
        "zig-out",
        ".zig-cache",
        "__pycache__",
        ".venv",
        "venv",
        "build",
        "dist",
        "target",
    }

    # Native Zig test conventions. Directory names are matched as whole path
    # segments; filenames are matched by anchored stem prefix/suffix. Matching
    # bare "test"/"spec" as a substring (the prior behaviour) misclassified
    # ordinary names like ``latest``/``contest``/``fastest.zig`` as tests.
    TEST_DIR_NAMES = {"test", "tests", "spec", "specs"}
    TEST_FILE_PREFIXES = ("test_", "spec_")
    TEST_FILE_SUFFIXES = ("_test.zig", "_spec.zig")

    def __init__(
        self,
        repo_path: str,
        skip_tests: bool = False,
        exclude_patterns: Optional[List[str]] = None,
    ):
        self.repo_path = Path(repo_path).resolve()
        self.skip_tests = skip_tests
        self.exclude_patterns = exclude_patterns or []

    def scan(self) -> Dict[str, Any]:
        """
        Scan the repository for Zig files.

        Returns scan_results.json structure:
        {
            "repository": "/path/to/repo",
            "scan_time": "2025-01-15T10:30:00",
            "files": [{"path": "src/main.zig", "size": 1234}, ...],
            "statistics": {...}
        }
        """
        files = []
        directories_scanned = 0
        directories_excluded = 0

        # followlinks=True so .zig files reachable ONLY through a symlinked
        # directory are still scanned (matches the iterdir()+is_dir() behaviour
        # of the other four language scanners). Following symlinks needs a guard
        # against two hazards, applied by pruning directory symlinks from `dirs`
        # (os.walk descends whatever remains in `dirs`):
        #   1. Alias duplication / canonical loss: a symlink pointing back INSIDE
        #      the repo names a real directory that os.walk already reaches on its
        #      canonical (non-symlink) path. Descending the alias would emit files
        #      under the symlink path and, worse, could suppress the real path
        #      entirely. We always drop the alias and keep the canonical visit.
        #   2. External symlink loops: a symlink pointing OUTSIDE the repo whose
        #      real target we have already descended (a self-referential loop)
        #      would recurse forever; an inode guard prevents re-descending it.
        # Traversal delegated to core/repo_walk.py. The previous implementation
        # used os.walk(followlinks=True) with an inode set that deduped only
        # REPEAT visits, so an external directory symlink was descended the first
        # time — `escape -> /` walked the host filesystem into the dataset. Both a
        # grep-based audit census and a token-matching parity test scored this
        # scanner "guarded"; only executing it against a hostile fixture found the
        # hole. It also never recorded unreadable directories, so a subtree it
        # could not enter vanished from a scan that reported success.
        stats: dict = {}

        def _on_file(entry, relative_path: str) -> None:
            # Case-insensitive: filesystems (macOS/Windows) and users may spell the
            # extension .ZIG/.Zig; skipping those silently loses files.
            if not entry.name.lower().endswith(".zig"):
                return
            if self.skip_tests and self._is_test_file(relative_path):
                return
            try:
                size = entry.stat().st_size
            except OSError:
                size = 0
            files.append({"path": relative_path, "size": size})

        def _should_exclude(name: str) -> bool:
            return (name in self.EXCLUDE_DIRS
                    or self._matches_exclude_pattern(name)
                    or (self.skip_tests and self._is_test_directory(name)))

        walk_repository(
            self.repo_path,
            should_exclude_directory=_should_exclude,
            on_file=_on_file,
            stats=stats,
        )
        directories_scanned = stats.get("directories_scanned", 0)
        directories_excluded = stats.get("directories_excluded", 0)
        self.walk_stats = stats

        total_size = sum(f["size"] for f in files)

        return {
            "repository": str(self.repo_path),
            "scan_time": datetime.now().isoformat(),
            "files": files,
            "statistics": {
                "total_files": len(files),
                "total_size_bytes": total_size,
                "directories_scanned": directories_scanned,
                "directories_excluded": directories_excluded,
                # Surfaced into the RESULT, not just stderr. A directory the
                # scanner could not enter is code it never analysed, and CI
                # discards stderr — so a gap that lives only in a warning is a
                # gap nobody sees.
                "directories_unreadable": stats.get("directories_unreadable", 0),
                "unreadable_examples": stats.get("unreadable_examples", []),
                # Symlinks are refused by policy; surfacing the count keeps that
                # a visible coverage gap rather than a silent false negative.
                "symlinks_skipped": stats.get("symlinks_skipped", 0),
                "symlink_examples": stats.get("symlink_examples", []),
            },
        }

    def _matches_exclude_pattern(self, name: str) -> bool:
        """Check if a name matches any exclude pattern (whole-segment match)."""
        return should_exclude_directory(name, self.exclude_patterns)

    def _is_test_directory(self, dirname: str) -> bool:
        """Check if a directory name indicates test code.

        Exact (whole-name) match so that ``latest``/``contest``/``attestation``
        are not misclassified as test directories.
        """
        return dirname.lower() in self.TEST_DIR_NAMES

    def _is_test_file(self, filepath: str) -> bool:
        """Check if a file path indicates test code.

        A file is a test iff one of its directory components is a test
        directory, or its filename is anchored (stem prefix ``test_``/``spec_``
        or suffix ``_test.zig``/``_spec.zig``). Anchoring stops ordinary names
        like ``src/fastest.zig``/``src/latest/main.zig`` from matching.
        """
        p = Path(filepath.lower())
        if any(part in self.TEST_DIR_NAMES for part in p.parts[:-1]):
            return True
        name = p.name
        if name.startswith(self.TEST_FILE_PREFIXES):
            return True
        if name.endswith(self.TEST_FILE_SUFFIXES):
            return True
        return False

    def save_results(self, output_path: str, results: Dict[str, Any]) -> None:
        """Save scan results to a JSON file."""
        write_json(output_path, results)

File: libs/vulnfounder-core/parsers/javascript/repository_scanner.js
#!/usr/bin/env node
/**
 * Repository Scanner
 *
 * Enumerates ALL source files in a repository for complete coverage.
 * This is Phase 1 of the parser upgrade to achieve full repository coverage.
 *
 * Usage:
 *   node repository_scanner.js <repo_path> [--output <file>] [--exclude <patterns>]
 *
 * Output (JSON):
 *   {
 *     "repository": "/path/to/repo",
 *     "scan_time": "2025-12-23T...",
 *     "files": [
 *       { "path": "relative/path/to/file.ts", "size": 1234, "extension": ".ts" }
 *     ],
 *     "statistics": {
 *       "total_files": 150,
 *       "by_extension": { ".ts": 100, ".js": 50 },
 *       "total_size_bytes": 500000,
 *       "directories_scanned": 25,
 *       "directories_excluded": 10
 *     }
 *   }
 */

const fs = require('fs');
const path = require('path');

class RepositoryScanner {
    constructor(repoPath, options = {}) {
        this.repoPath = path.resolve(repoPath);
        this.skipTests = options.skipTests || false;

        // Default exclude patterns
        this.excludePatterns = options.excludePatterns || [
            'node_modules',
            'dist',
            'build',
            'coverage',
            '.git',
            '.svn',
            '.hg',
            '__pycache__',
            '.next',
            '.nuxt',
            'out',
            '.cache',
            'tmp',
            'temp',
            '.turbo',
            '.vercel',
            '.netlify'
        ];

        // Source file extensions to include
        this.sourceExtensions = options.sourceExtensions || [
            '.js',
            '.ts',
            '.jsx',
            '.tsx',
            '.mjs',
            '.cjs'
        ];

        // Statistics
        this.stats = {
            totalFiles: 0,
            byExtension: {},
            totalSizeBytes: 0,
            directoriesScanned: 0,
            directoriesExcluded: 0,
            testFilesSkipped: 0,
            // snake_case coverage keys, matching the Python-family parsers so the
            // aggregator's presence probe (core/scanner.py) reads them. Present
            // at 0 marks this parser as coverage-instrumented (vs. a parser that
            // reports no coverage at all, where the key is absent).
            symlinks_skipped: 0,
            symlink_examples: [],
            directories_unreadable: 0,
            unreadable_examples: []
        };

        // Results
        this.files = [];
    }

    /**
     * Check if a file is a test file
     */
    isTestFile(relativePath) {
        const base = path.basename(relativePath);
        const dir = relativePath.replace(/\\/g, '/');

        // Directory-based patterns
        if (dir.includes('__tests__/') || dir.includes('__mocks__/') ||
            /(?:^|\/)tests?\//.test(dir)) {
            return true;
        }

        // File name patterns: .test.*, .spec.*, _test.*, test_*
        if (/\.(test|spec)\.[jt]sx?$/.test(base)) return true;
        if (/_(test)\.[jt]sx?$/.test(base)) return true;
        if (/^test_/.test(base)) return true;

        return false;
    }

    /**
     * Check if a directory should be excluded
     */
    shouldExcludeDirectory(dirName) {
        return this.excludePatterns.some(pattern => {
            // Exact match
            if (dirName === pattern) return true;
            // Glob-like match (e.g., pattern ends with *)
            if (pattern.endsWith('*') && dirName.startsWith(pattern.slice(0, -1))) return true;
            return false;
        });
    }

    /**
     * Check if a file is a source file we want to include
     */
    isSourceFile(fileName) {
        const ext = path.extname(fileName).toLowerCase();
        return this.sourceExtensions.includes(ext);
    }

    /**
     * Recursively scan a directory
     */
    scanDirectory(dirPath, relativePath = '') {
        this.stats.directoriesScanned++;

        let entries;
        try {
            entries = fs.readdirSync(dirPath, { withFileTypes: true });
        } catch (error) {
            // Unreadable directory: record it as a counted coverage gap rather
            // than a silent skip (a silent skip is a false negative, the worst
            // direction for a SAST tool), then stop descending this one.
            this.stats.directories_unreadable++;
            if (this.stats.unreadable_examples.length < 5) {
                this.stats.unreadable_examples.push(`${relativePath || '.'}: ${error.message}`);
            }
            console.error(`Warning: Cannot read directory ${dirPath}: ${error.message}`);
            return;
        }

        for (const entry of entries) {
            const fullPath = path.join(dirPath, entry.name);
            const entryRelativePath = relativePath ? path.join(relativePath, entry.name) : entry.name;

            // Refuse symlinked directories. The scanned repository is untrusted:
            // `vendor -> /` would walk the host filesystem into the dataset (and
            // from there to the model provider), and `loop -> ..` never terminates.
            //
            // This check is currently redundant — readdirSync({withFileTypes:true})
            // returns Dirents with lstat semantics, so a symlink reports
            // isSymbolicLink() and isDirectory() is already false. It is written out
            // anyway because that safety is an undocumented property of the API
            // choice, not of this code: switching to statSync, or to a readdir
            // without withFileTypes, would silently restore the vulnerability with
            // no visible diff at this line. The sibling Python scanners had exactly
            // this bug.
            if (entry.isSymbolicLink()) {
                // Count as a symlink skip, NOT directoriesExcluded — that field
                // is for excluded real directories, and folding refusals into it
                // both mislabels the gap and hides it from the coverage probe,
                // which reads snake_case symlinks_skipped.
                this.stats.symlinks_skipped++;
                if (this.stats.symlink_examples.length < 5) {
                    this.stats.symlink_examples.push(entryRelativePath);
                }
                continue;
            }

            if (entry.isDirectory()) {
                if (this.shouldExcludeDirectory(entry.name)) {
                    this.stats.directoriesExcluded++;
                    continue;
                }
                // Recurse into subdirectory
                this.scanDirectory(fullPath, entryRelativePath);
            } else if (entry.isFile()) {
                if (this.isSourceFile(entry.name)) {
                    // Skip test files if requested
                    if (this.skipTests && this.isTestFile(entryRelativePath)) {
                        this.stats.testFilesSkipped++;
                        continue;
                    }

                    let fileStats;
                    try {
                        fileStats = fs.statSync(fullPath);
                    } catch (error) {
                        console.error(`Warning: Cannot stat file ${fullPath}: ${error.message}`);
                        continue;
                    }

                    const ext = path.extname(entry.name).toLowerCase();

                    this.files.push({
                        path: entryRelativePath,
                        size: fileStats.size,
                        extension: ext
                    });

                    // Update statistics
                    this.stats.totalFiles++;
                    this.stats.totalSizeBytes += fileStats.size;
                    this.stats.byExtension[ext] = (this.stats.byExtension[ext] || 0) + 1;
                }
            }
        }
    }

    /**
     * Run the scan and return results
     */
    scan() {
        if (!fs.existsSync(this.repoPath)) {
            throw new Error(`Repository path does not exist: ${this.repoPath}`);
        }

        if (!fs.statSync(this.repoPath).isDirectory()) {
            throw new Error(`Repository path is not a directory: ${this.repoPath}`);
        }

        this.files = [];
        this.stats = {
            totalFiles: 0,
            byExtension: {},
            totalSizeBytes: 0,
            directoriesScanned: 0,
            directoriesExcluded: 0,
            testFilesSkipped: 0,
            // snake_case coverage keys, matching the Python-family parsers so the
            // aggregator's presence probe (core/scanner.py) reads them. Present
            // at 0 marks this parser as coverage-instrumented (vs. a parser that
            // reports no coverage at all, where the key is absent).
            symlinks_skipped: 0,
            symlink_examples: [],
            directories_unreadable: 0,
            unreadable_examples: []
        };

        this.scanDirectory(this.repoPath);

        // Sort files by path for consistent output
        this.files.sort((a, b) => a.path.localeCompare(b.path));

        return {
            repository: this.repoPath,
            scan_time: new Date().toISOString(),
            files: this.files,
            statistics: this.stats
        };
    }
}

// CLI interface
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        console.error('Usage: node repository_scanner.js <repo_path> [--output <file>] [--exclude <pattern1,pattern2,...>]');
        console.error('');
        console.error('Options:');
        console.error('  --output <file>     Write results to file instead of stdout');
        console.error('  --exclude <patterns> Additional comma-separated exclude patterns');
        console.error('');
        console.error('Example:');
        console.error('  node repository_scanner.js /path/to/repo --output scan_results.json');
        process.exit(1);
    }

    const repoPath = args[0];
    let outputFile = null;
    let additionalExcludes = [];
    let skipTests = false;

    // Parse arguments
    for (let i = 1; i < args.length; i++) {
        if (args[i] === '--output' && args[i + 1]) {
            outputFile = args[i + 1];
            i++;
        } else if (args[i] === '--exclude' && args[i + 1]) {
            additionalExcludes = args[i + 1].split(',').map(s => s.trim());
            i++;
        } else if (args[i] === '--skip-tests') {
            skipTests = true;
        }
    }

    try {
        const options = {};
        if (skipTests) {
            options.skipTests = true;
        }
        if (additionalExcludes.length > 0) {
            // Merge with default excludes
            const defaultExcludes = [
                'node_modules', 'dist', 'build', 'coverage', '.git', '.svn', '.hg',
                '__pycache__', '.next', '.nuxt', 'out', '.cache', 'tmp', 'temp',
                '.turbo', '.vercel', '.netlify'
            ];
            options.excludePatterns = [...defaultExcludes, ...additionalExcludes];
        }

        const scanner = new RepositoryScanner(repoPath, options);
        const result = scanner.scan();

        const output = JSON.stringify(result, null, 2);

        if (outputFile) {
            fs.writeFileSync(outputFile, output);
            console.error(`Scan complete. Results written to: ${outputFile}`);
            console.error(`Total files found: ${result.statistics.totalFiles}`);
            console.error(`By extension:`, result.statistics.byExtension);
        } else {
            console.log(output);
        }

        process.exit(0);
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

module.exports = { RepositoryScanner };
